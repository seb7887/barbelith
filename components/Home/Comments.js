import React from 'react';
import Link from 'next/link';
import CardHeader from "@material-ui/core/CardHeader";
import FormControl from "@material-ui/core/FormControl";
import InputLabel from "@material-ui/core/InputLabel";
import Input from "@material-ui/core/Input";
import Avatar from "@material-ui/core/Avatar";
import Delete from "@material-ui/icons/Delete";
import withStyles from "@material-ui/core/styles/withStyles";

class Comments extends React.Component {
  state = {
    text: ''
  };

  handleChange = (e) => {
    e.preventDefault();
    this.setState({ text: e.target.value });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { text } = this.state;
    const { postId, handleAddComment } = this.props;

    handleAddComment(postId, text);
    this.setState({ text: '' });
  }

  showComment = (comment) => {
    const { postId, auth, classes, handleDeleteComment } = this.props;
    const isCommentCreator = comment.postedBy._id === auth.user._id;

    return (
      <div data-testid='comment'>
        <Link href={`/profile/${comment.postedBy._id}`}>
          <a className={classes.link}>{comment.postedBy.name}</a>
        </Link>
        <br />
        {comment.text}
        <span className={classes.commentDate}>
          {comment.createdAt}
          {isCommentCreator && (
            <Delete
              data-testid='delete-comment'
              color='secondary'
              className={classes.commentDelete}
              onClick={() => handleDeleteComment(postId, comment)}
            />
          )}
        </span>
      </div>
    );
  }

  render() {
    const { auth, comments, classes } = this.props;
    const { text } = this.state;

    return (
      <div className={classes.comments}>
        {/* Comment Input */}
        <CardHeader
          avatar={
            <Avatar className={classes.smallAvatar} src={auth.user.avatar} />
          }
          title={
            <form onSubmit={this.handleSubmit} data-testid='form'>
              <FormControl margin='normal' required fullWidth>
                <InputLabel htmlFor='add-comment'>Add comments</InputLabel>
                <Input
                  id='add-comment'
                  name='text'
                  placeholder='Something to say?'
                  value={text}
                  onChange={this.handleChange}
                />
              </FormControl>
            </form>
          }
          className={classes.cardHeader}
        />

        {/* Comments */}
        {comments.map(comment => (
          <CardHeader
            key={comment._id}
            avatar={
              <Avatar
                className={classes.smallAvatar}
                src={comment.postedBy.avatar}
              />
            }
            title={this.showComment(comment)}
            className={classes.cardHeader}
          />
        ))}
      </div>
    );
  }
}

const styles = theme => ({
  comments: {
    backgroundColor: 'rgba(11, 61, 130, 0.06)'
  },
  cardHeader: {
    paddingTop: theme.spacing.unit,
    paddingBottom: theme.spacing.unit
  },
  smallAvatar: {
    margin: 10
  },
  commentDate: {
    display: 'block',
    color: 'gray',
    fontSize: '0.8em'
  },
  commentDelete: {
    fontSize: '1.6em',
    verticalAlign: 'middle',
    cursor: 'pointer'
  },
  link: {
    textDecoration: 'none',
    color: theme.palette.secondary.main,
  }
});

export default withStyles(styles)(Comments);
