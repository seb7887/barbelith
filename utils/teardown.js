use barbelith-db;
show collections;
db.users.remove({});
db.sessions.remove({});
db.posts.remove({});
