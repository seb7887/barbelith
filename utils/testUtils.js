import faker from 'faker';

export const validUser = () => ({
  name: 'kingmob',
  email: 'king@mob.com',
  password: 'km1234'
});

export const fakeUser = () => ({
  name: faker.internet.userName(),
  email: faker.internet.email(),
  password: faker.internet.password()
});

export const mockedUser = () => ({
  _id: '7',
  avatar: '/static/images/avatar.jpg',
  createdAt: '2019-02-1234',
  email: faker.internet.email(),
  followers: ['1', '2', '3'],
  following: ['3'],
  name: 'Test',
  updatedAt: '2019-02-1234'
});

export const updatedUser = () => ({
  _id: '7',
  name: 'kingmob',
  email: faker.internet.email(),
  avatar: faker.image.avatar(),
  about: faker.lorem.text()
});

export const generateUsers = quantity => {
  const generatedUsers = [];

  for (let i = 0; i < quantity; i++) {
    const user = {
      _id: `${i}`,
      name: `test${i}`,
      avatar: faker.image.avatar()
    };
    generatedUsers.push(user);
  }

  return generatedUsers;
};

export const generateComments = quantity => {
  const generatedComments = [];

  for (let i = 0; i < quantity; i++) {
    const comment = {
      _id: faker.random.uuid().toString(),
      postedBy: {
        _id: '7',
        name: 'test',
        avatar: faker.image.avatar()
      },
      text: faker.lorem.text()
    };
    generatedComments.push(comment);
  }
  return generatedComments;
};

export const fakePost = () => ({
  _id: '7',
  postedBy: {
    _id: '7',
    name: 'test',
    avatar: faker.image.avatar()
  },
  createdAt: faker.date.recent().toString(),
  text: faker.lorem.text(),
  image: faker.image.avatar(),
  likes: ['7'],
  comments: []
});
