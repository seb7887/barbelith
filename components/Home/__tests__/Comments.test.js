/* eslint-env jest */
import 'react-testing-library/cleanup-after-each';
import 'jest-dom/extend-expect';

import { render, fireEvent } from 'react-testing-library';
import faker from 'faker';
import { generateComments } from '../../../utils/testUtils';
import Comments from '../Comments';

const handleAddComment = jest.fn();
const handleDeleteComment = jest.fn();

let auth = {
  user: {
    _id: '7'
  }
}

const comments = generateComments(3);
const postId = '8';

describe('<Comments/>', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = render(
      <Comments
        auth={auth}
        postId={postId}
        comments={[]}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should display comments correctly', () => {
    const { getAllByTestId } = render(
      <Comments
        auth={auth}
        postId={postId}
        comments={comments}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    expect(getAllByTestId('comment')).toHaveLength(3);
  });

  it('should add a comment', () => {
    const { getByTestId, getByLabelText } = render(
      <Comments
        auth={auth}
        postId={postId}
        comments={comments}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    fireEvent.change(getByLabelText(/add comments/i), { target: { value: faker.lorem.text() } })

    const form = getByTestId('form');
    fireEvent.submit(form);

    expect(handleAddComment).toHaveBeenCalledTimes(1);
  });

  it('should delete a comment', () => {
    const { getByTestId } = render(
      <Comments
        auth={auth}
        postId={postId}
        comments={comments}
        handleAddComment={handleAddComment}
        handleDeleteComment={handleDeleteComment}
      />
    );

    fireEvent.click(getByTestId('delete-comment'));

    expect(handleDeleteComment).toHaveBeenCalledTimes(1);
  });
});
