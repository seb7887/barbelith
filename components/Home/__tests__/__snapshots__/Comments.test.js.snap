// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments/> renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Comments-comments-1"
  >
    <div
      class="MuiCardHeader-root-7 Comments-cardHeader-2"
    >
      <div
        class="MuiCardHeader-avatar-8"
      >
        <div
          class="MuiAvatar-root-13 MuiAvatar-colorDefault-14 Comments-smallAvatar-3"
        />
      </div>
      <div
        class="MuiCardHeader-content-10"
      >
        <span
          class="MuiTypography-root-16 MuiTypography-body2-24 MuiCardHeader-title-11"
        >
          <form
            data-testid="form"
          >
            <div
              class="MuiFormControl-root-52 MuiFormControl-marginNormal-53 MuiFormControl-fullWidth-55"
            >
              <label
                class="MuiFormLabel-root-67 MuiFormLabel-required-72 MuiInputLabel-required-60 MuiInputLabel-root-56 MuiInputLabel-formControl-61 MuiInputLabel-animated-64"
                data-shrink="false"
                for="add-comment"
              >
                Add comments
                <span
                  class="MuiFormLabel-asterisk-73"
                >
                  â€‰*
                </span>
              </label>
              <div
                class="MuiInputBase-root-87 MuiInput-root-74 MuiInput-underline-78 MuiInputBase-formControl-88 MuiInput-formControl-75"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input-97 MuiInput-input-82"
                  id="add-comment"
                  name="text"
                  placeholder="Something to say?"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
          </form>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
